Add Mistral Provider Support

rajkstats opened the pull request:
## Changes
- Added MistralBaseProvider class to llm_provider.py for better code organization
-  MistralProvider to inherit from MistralBaseProvider
- Fixed tool schema handling for Mistral API compatibility
- Added support for Pixtral (vision) and Mistral Large models

## Testing
- Tested tool calling functionality with Mistral API
- Verified vision capabilities with Pixtral model
- Confirmed compatibility with existing agent infrastructure

## Notes
- Requires MISTRAL_API_KEY and HF_TOKEN in .env file
- Pixtral model supports both vision and tool calls
- Mistral Large model used for non-vision tasks

jamesmurdza commented:
Thanks, Raj.

Pixtral and Mistral Large seem like the right combination.

What's the reason for adding MistralBaseProvider? I checked and the implementation looks the same as OpenAIBaseProvider.

Also, HF_TOKEN isn't required.

rajkstats commented:
@jamesmurdza Thanks for the feedback! 

You're right about both points:

1. Removed MistralBaseProvider since Mistral's API follows OpenAI's spec - MistralProvider now inherits directly from OpenAIBaseProvider.

2. Removed HF_TOKEN from OSAtlasProvider. I initially added it while debugging gradio rate limits, but it's not needed for basic functionality.

The PR is now focused on:
- Adding Mistral support through OpenAIBaseProvider
- Adding Pixtral (vision) and Mistral Large (actions) models
- Updating README with Mistral documentation

Let me know if you'd like any other changes!

jamesmurdza commented:
Thanks for clarifying that!

I actually added the HF_TOKEN for that reason, but I could never tell if it made any difference on the rate limit. Were you able to tell? If it doesn't do anything, we might remove it then.

James

rajkstats commented:
@jamesmurdza I noticed that without the token, OS Atlas eventually stops responding after a series of requests. So if you're planning on heavy testing, you’ll probably need it. For now, I've removed it from the PR.

rajkstats commented:
Re-added the optional HF token parameter in the OSAtlasProvider initialization. This change enables authenticated Gradio requests to mitigate rate limits when necessary.

jamesmurdza commented:
Functionally it seems fine. Can you please look through all the changes that are in this PR and keep the ones that are really necessary? There are still many changes that are unexplained.

jamesmurdza commented on os_computer_use/llm_provider.py:
```diff
@@ -49,7 +49,10 @@ def create_function_schema(self, definitions):
             required = []
 
             for param_name, param_desc in details["params"].items():
-                properties[param_name] = {"type": "string", "description": param_desc}
```
This

rajkstats replied:
I added this check to ensure that the tool schema matches what Mistral expects. Without this, when the parameter description is given as a plain string it gets interpreted as a string literal rather than a dict with "type" and "description", leading to a validation error (Invalid tool schema: {'type': 'string', 'description': 'Shell command to run synchronously'} is not of type 'string'). This change wraps non-dict values properly, ensuring compatibility with Mistral's schema requirements.

jamesmurdza commented on os_computer_use/providers.py:
```diff
@@ -3,6 +3,9 @@
 from os_computer_use.llm_provider import OpenAIBaseProvider, AnthropicBaseProvider
 from os_computer_use.osatlas_provider import OSAtlasProvider
 from os_computer_use.showui_provider import ShowUIProvider
+from openai import OpenAI
```
This

rajkstats replied:
Removing this, not required

jamesmurdza commented on os_computer_use/sandbox_agent.py:
```diff
@@ -53,9 +53,14 @@ def call_function(self, name, arguments):
 
     def tool(description, params):
         def decorator(func):
-            tools[func.__name__] = {"description": description, "params": params}
```
This

jamesmurdza replied:
value should always be a string

jamesmurdza commented on os_computer_use/sandbox_agent.py:
```diff
@@ -176,6 +181,8 @@ def append_screenshot(self):
         )
 
     def run(self, instruction):
+        # ensure the last message is from the user (not assistant)
```
This

jamesmurdza commented on os_computer_use/sandbox_agent.py:
```diff
@@ -193,10 +200,12 @@ def run(self, instruction):
                     *self.messages,
                     Message(
-                        logger.log(f"THOUGHT: {self.append_screenshot()}", "green")
+                        logger.log(f"THOUGHT: {self.append_screenshot()}", "green"),
```
This

jamesmurdza commented on os_computer_use/sandbox_agent.py:
```diff
@@ -193,10 +200,12 @@ def run(self, instruction):
                     ),
                     Message(
                         "I will now use tool calls to take these actions, or use the stop command if the objective is complete.",
+                        role="user"
```
This

rajkstats replied:
I adjusted the messaging order so that the final message in the sequence has the role "user"—Mistral's serving rules demand that the last role is either "user" or "tool" (or an assistant message with the appropriate True prefix). Without this, the assistant's message at the end causes a schema validation error ("Expected last role User or Tool (or Assistant with prefix True) for serving but got assistant"). This change is necessary to satisfy the expectations enforced by Mistral.

jamesmurdza replied:
Look at this: https://docs.mistral.ai/guides/prefix/

jamesmurdza replied:
The problem with just changing "assistant" to "user" is that this is intended to be an assistant prompt.

jamesmurdza replied:
The prefix option is what you're looking for.

rajkstats replied:
Thanks for sharing @jamesmurdza , I've incorporated this now

jamesmurdza commented on tests/llm_provider.py:
```diff
@@ -3,14 +3,20 @@
     "click_item": {
         "description": "Click on an item on the screen",
-        "params": {"description": "Description of the item to click on"},
+        "params": {
```
This

rajkstats replied:
Keeping only necessary changes

jamesmurdza reviewed (commented):
I'm referring to these changes.

rajkstats commented:
@jamesmurdza I've resolved all reported issues while retaining only the necessary changes. The tool schema check and messaging order adjustments ensure proper validation and functionality. Let me know if you see any further concerns.

jamesmurdza commented:
@rajkstats Please show me that you understand what the new code does.

jamesmurdza commented:
For each code change we need to see:
1) How it's connected to Mistral
2) That it doesn't break the existing support for any of the **other providers**

jamesmurdza commented:
@rajkstats What do you think about implementing it like this instead?

```python
class MistralBaseProvider(OpenAIBaseProvider):

    def create_function_def(self, name, details, properties, required):
        # Wrap all descriptions to avoid Mistral validation error.
        details["description"] = {"type": "string", "description": details.get("description")}
        return super().create_function_def(name, details, properties, required)
    
    def call(self, messages, functions=None):
        # Make the last message a prefix if its role is assistant.
        if messages and messages[-1].get("role") == "assistant":
            messages[-1]["prefix"] = True
        return super().call(messages, functions)
```

The idea behind the base provider classes is to handle discrepancies between how different providers handle tool calls and messages.

rajkstats commented:
@jamesmurdza Apologies for the confusion here, detailed comments were in the conversation. 

Based on your feedback, now I've implemented and tested the following:
- Added a new MistralProvider in providers with support for "small", "medium", "large", and "pixtral".
- Updated sandbox_agent to conditionally add "prefix": True for the detailed screenshot prompt when using Mistral.
- For non-Mistral providers, the prompt is sent as a normal user message to avoid extra input errors.

Apart from this I've introduced HF token and couple of Readme changes for Mistral. 

Hope this makes sense now.

rajkstats commented:
@jamesmurdza I like the idea, it’s cleaner and aligns with the docs. 

However, in our testing the simple approach (just marking the assistant message with "prefix": True) didn’t work

There are two issues with this approach: 

**Description Issue**: You are wrapping the description in a dict

```python   
details["description"] = {"type": "string", "description": details.get("description")}
```
but Mistral expects it to be a plain string like:
 
```python  
"Run a shell command and return the result."
```

we should ensure that if description is a dict, we extract the inner string:

```python   
   if isinstance(details.get("description"), dict):
       details["description"] = details["description"].get("description", "")
```       

**Call Method Behavior**

Following sets a prefix flag on the assistant message:

```python  
   if messages and messages[-1].get("role") == "assistant":
       messages[-1]["prefix"] = True
```       

but it doesn’t merge the assistant’s output with the conversation

For example: We have following in our conversation
```python  
messages = [
       {"role": "user", "content": "search for cats on internet"},
       {"role": "assistant", "content": "click on the web browser icon"}
   ]
```

When call method runs, it becomes 

```python  
messages =
   [
       {"role": "user", "content": "search for cats on internet"},
       {"role": "assistant", "content": "click on the web browser icon", "prefix": True}
   ]
```   
The assistant’s content ("click on the web browser icon") isn’t fed back into the conversation—it just sits there with a flag, and the tools never see that instruction as part of a new user input.

Now, I'm merging the assistant response

```python  
   if messages and messages[-1].get("role") == "assistant":
       prefix = messages.pop()["content"]
       if messages and messages[-1].get("role") == "user":
           messages[-1]["content"] = prefix + "\n" + messages[-1].get("content", "")
       else:
           messages.append({"role": "user", "content": prefix})
```

This means the final message becomes:

```python
  [
       {"role": "user", "content": "click on the web browser icon\nsearch for cats on internet"}
   ]
```

Now the assistant’s guidance is merged with the user’s original command, ensuring the context is preserved and the loop continues.  Hope this makes sense.

rajkstats commented:
Summary of Changes:

- **llm_provider.py**: Updated the MistralBaseProvider to properly unwrap descriptions and merge the assistant's output with the conversation.
- **config.py & README**: Adjusted configuration settings and documentation to reflect the new Mistral integrations.
- **os_computer_use/osatlas_provider.py**: Added HF_token support.
- **os_computer_use/providers.py**: Introduced the new MistralProvider.
- **tests/llm_provider.py**: Added tests specific to Mistral changes.

jamesmurdza commented:
It looks clean. I'll test it and then merge.
